_posts/2016/2016-09-28-rest-api-test-framework.md

```diff
@@ -0,0 +1,191 @@
+---
+layout: post
+title: "How we test our API"
+date: "2016-09-28 11:11:11"
+image: blog-header/heartbeat.jpg
+categories: tech-stories
+authors: ["Ulf B."]
+---
+
+A few years ago, at the time when the development of the ePages REST API started,
+ePages 6 was just a huge monolith written in Perl and already hard to handle.
+Since the monolith should not be inflated even more and in order to be able to use state-of-the-art frameworks,
+the REST API was started as a separate service implemented in Java.
+
+This service communicates with the monolith via REST and SOAP in order to operate on the data we won't to expose to the public.
+All the API related business logic like right management and rate limits are implemented in Java such that the API's we had to implement in Perl could
+become quiete simple.
+
+Of course all the resource and services classes are coverey by unit test on both sides, Perl and Java,
+but since the functionality of the API is based on two different projects that live in two different repositories
+which have two different teams responsible for the releases,
+we wanted to have a simple way to check with one click if everything - from end to end - is working as expected.
+
+This was the starting point of RAT - our Rest API Test framework.
+
+## RAT
+
+The idea of RAT was to create a framework that allows us to easily verify the state of the REST API on an abritrary ePages installation.
+
+In order to do so, RAT sends requests to the API and validates the response.
+To be able to compare the response with an expected one, we let the test run on shop of the DemoShopType
+(The developer installation of ePages includes a DemoShopType which allows to create a demo shop with certain products of different types and a few orders).
+We wanted the test cases to be able to run in a arbitrary order. The problem is that some of the API calls change a shop (DELETE product),
+which lead to different results in the following calls. The solution is to create new shops for testing those calls.
+The ePages6 REST API does not offer an endpoint to create new shops so far, so we used the ePages SOAP API to do that.
+
+RAT is based on [REST-assured](http://rest-assured.io/) and [Serenity BDD](http://www.thucydides.info), two tools that take over a lot of work on the way to produce nice test results. In the next paragraphs I will introduce these tools briefly and show how we use them.
```

Comment:
ahh..Das hatte ich falsch verstanden. Dann würde man sagen: "RAT is based on REST-assured and Serenity BDD, two tools that ease a lot of our workload and achieve good test results. In the following paragraphs, I will introduce the mentioned tools briefly and show how we use them."